styx-common/src/main/java/com/spotify/styx/api/Middlewares.java

```diff
@@ -220,6 +226,33 @@ static GoogleIdToken verifyIdToken(String s, GoogleIdTokenVerifier verifier) {
     };
   }
 
+  public static <T> Middleware<AsyncHandler<Response<T>>, AsyncHandler<Response<T>>> tracer(
+      Tracer tracer, String service) {
+    return innerHandler -> requestContext -> {
+      final Request request = requestContext.request();
+      final URI uri = URI.create(request.uri());
+      final Span span = tracer.spanBuilder(service + '/' + uri.getPath()).startSpan();
+      span.putAttribute("method", stringAttributeValue(request.method()));
+      span.putAttribute("uri", stringAttributeValue(request.uri()));
+      final CompletionStage<Response<T>> response;
+      try  {
+        response = innerHandler.invoke(requestContext);
+      } catch (Exception e) {
+        span.setStatus(UNKNOWN);
+        span.end();
+        Throwables.throwIfUnchecked(e);
+        throw new RuntimeException(e);
+      }
+      response.whenComplete((r, ex) -> {
+        if (ex != null) {
+          span.setStatus(UNKNOWN);
+        }
+        span.end();
```

Comment:
Doing `end()` in multiple places. Consider try-with-resource instead? https://static.javadoc.io/io.opencensus/opencensus-api/0.13.2/io/opencensus/trace/SpanBuilder.html#startScopedSpan--

Reply:
The request handler might be async, in which case it returns a future and we'd end up closing the span before it's (asynchronously) completed.